plugin/meshnet.go

```diff
@@ -245,76 +247,80 @@ func cmdAdd(args *skel.CmdArgs) error {
-					log.Infof("Only local link exists, removing it first")
+					log.Infof("Only local link exists, removing it first local pod %s and peer pod %s", localPod.Name, peerPod.Name)
 					if err := myVeth.RemoveVethLink(); err != nil {
 						log.Errorf("Failed to remove a local stale VEth interface %s for pod %s", myVeth.LinkName, localPod.Name)
 						return err
 					}
-					log.Infof("Adding the new veth link to both pods")
+					log.Infof("Adding the new veth link to both pods local pod %s and peer pod %s", localPod.Name, peerPod.Name)
 					if err = koko.MakeVeth(*myVeth, *peerVeth); err != nil {
-						log.Errorf("Error creating VEth pair after local link remove: %s", err)
+						log.Errorf("Error creating VEth pair after local link remove: %s, local pod %s and peer pod %s", err, localPod.Name, peerPod.Name)
 						return err
 					}
 				} else { // if neither link exists, we have two options
-					log.Infof("Neither link exists. Checking if we've been skipped")
+					log.Infof("Neither link exists. Checking if we've been skipped local pod %s and peer pod %s", localPod.Name, peerPod.Name)
 					isSkipped, err := meshnetClient.IsSkipped(ctx, &mpb.SkipQuery{
 						Pod:    localPod.Name,
 						Peer:   peerPod.Name,
 						KubeNs: string(cniArgs.K8S_POD_NAMESPACE),
 					})
 					if err != nil {
-						log.Errorf("Failed to read skipped status from our peer")
+						log.Errorf("Local pod %s Failed to read skipped status from our peer %s", localPod.Name, peerPod.Name)
 						return err
 					}
-					log.Infof("Have we been skipped by our peer %s? %t", peerPod.Name, isSkipped.Response)
+					log.Infof("Have we %s been skipped by our peer %s? %t", localPod.Name, peerPod.Name, isSkipped.Response)
 
 					// Comparing names to determine higher priority
 					higherPrio := localPod.Name > peerPod.Name
-					log.Infof("DO we have a higher priority? %t", higherPrio)
+					log.Infof("DO we %s have a higher priority than peer %s ? %t", localPod.Name, peerPod.Name, higherPrio)
 
 					if isSkipped.Response || higherPrio { // If peer POD skipped us (booted before us) or we have a higher priority
-						log.Infof("Peer POD has skipped us or we have a higher priority")
+						log.Infof("Peer POD %s has skipped us or we %s have a higher priority", peerPod.Name, localPod.Name)
 						if err = koko.MakeVeth(*myVeth, *peerVeth); err != nil {
 							log.Errorf("Error when creating a new VEth pair with koko: %s", err)
-							log.Infof("MY VETH STRUCT: %+v", spew.Sdump(myVeth))
-							log.Infof("PEER STRUCT: %+v", spew.Sdump(peerVeth))
-							return err
+							log.Infof("local pod %s and peer pod %s MY VETH STRUCT: %+v", localPod.Name, peerPod.Name, spew.Sdump(myVeth))
+							log.Infof("local pod %s and peer pod %s PEER STRUCT: %+v", localPod.Name, peerPod.Name, spew.Sdump(peerVeth))
+							if strings.Contains(err.Error(), "file exists") {
+								log.Infof("race condition hit local pod %s and peer pod %s", localPod.Name, peerPod.Name)
```

Comment:
My thought was the peer pod which is creating the link in first place ( probably because it has higher priority), will handle any failures.

Also question for you:
can you elaborate the skipping logic, 
1. can we assume if we have higher prio, then we create link ?
2.  Is isSkipped.Response == True when both pods are up at same time or one of them skips based on priority ?

We could handle this condition probably differently based on above

if isSkipped.Response || higherPrio { // If peer POD skipped us (booted before us) or we have a higher priority

Reply:
yeah, I guess if the peer pod fails to create the link then the error would bubble up and CNI will retry.


wrt your question, the skipping logic is a bit hard to reason about, mainly because of the CNI deletion command (when CNI fails to correctly setup the interfaces or Pod gets deleted by kubelet). 

When Pods are being created for the first time, skipped is set by a pod on its peer if the peer is not alive yet. the idea is that once the peer comes up, it will plug in all of the interfaces. So when `isSkipped.Response == True`, we **have** to do the work. Priority is a local tie-breaker for the condition when two pods are both coming up at the same time and none of them is skipped. 

So, in your case, I'd expect that the existing interface should be detected by this logic https://github.com/networkop/meshnet-cni/blob/56abbfc48205285cd2b18d5aaf78f69b56c0a7b6/plugin/meshnet.go#L261-L262

I still don't fully understand what needs to happen for a pod to come up, reach this stage of the code and see this error.  I'm fine with this solution as a workaround but ideally, I'd prefer to handle this explicitly in the code logic rather than catching an error.

Reply:
Based on your logs, it's clear that the peer pod has come up before us and has skipped us, so priority doesn't come into play, only the `skipped` flag is true. Which means that the peer pod should have never even attempted to plug in the interfaces, which is why I don't understand why you're getting the `file exists` error.
```
time="2023-04-26T21:33:42-04:00" level=info msg="Creating Veth struct with NetNS:/proc/19196/ns/net and intfName: eth5, IP:"
time="2023-04-26T21:33:42-04:00" level=info msg="Does the link already exist? Local:false, Peer:false"
time="2023-04-26T21:33:42-04:00" level=info msg="Neither link exists. Checking if we've been skipped"
time="2023-04-26T21:33:43-04:00" level=info msg="Have we been skipped by our peer test? true"
time="2023-04-26T21:33:43-04:00" level=info msg="DO we have a higher priority? false"
time="2023-04-26T21:33:43-04:00" level=info msg="Peer POD has skipped us or we have a higher priority"
time="2023-04-26T21:33:43-04:00" level=error msg="Error when creating a new VEth pair with koko: failed to rename link koko1662009491 -> e1-21: file exists"
```

Reply:
@sar772004, I tried to reproduce this issue with a full mesh topology of 8 pods in a kind cluster with a single worker node. But I am not able to reproduce this issue in my setup. It does not happen during new topology creation. I tried to delete/add/replace a pod and I get a different issue which is expected but I dont see this issue. Can you please share following information? 
- steps (along with commands) to reproduce this issue
- topology file if possible

Reply:
@networkop  I understand, ideally it would be better to get a fix which does not try to create an already existing interface. Please see my response below for logs with extra debugging which may help you understand the issue better.

@manomugdha 
 in general, my observations 
1.  out of 8 containers, 1 of them called "test" comes first, rest of the containers from dut-a to dut-f and dut-bridge, come a little slower due to slower startup script.
2. i have attached the logs with my changes which show clearly which peer and local pod is being processed and sequence of events. Its fairly clear that there is a race condition in the cmdAdd.
[meshnet-cni.log](https://github.com/networkop/meshnet-cni/files/11369152/meshnet-cni.log)

topology file 
[meshnet_links.yaml.txt](https://github.com/networkop/meshnet-cni/files/11369159/meshnet_links.yaml.txt)

Reply:
I was just going through the logs when I saw this message.... excellent analysis, now it makes sense. 
I agree, the skipping logic does not take into account the fact of multiple links but it should. 
There also seems to be another bug that I've stumbled upon accidentally. Seems like meshnet would not plug in the loopback links. can you confirm this @sar772004 ?

```
time="2023-05-01T20:27:15-04:00" level=info msg="Have we dut-c been skipped by our peer dut-c? false"
time="2023-05-01T20:27:15-04:00" level=info msg="DO we dut-c have a higher priority than peer dut-c ? false"
time="2023-05-01T20:27:15-04:00" level=info msg="Doing nothing, expecting peer pod dut-c to connect veth pair with local pod dut-c"
```

Reply:
@networkop you the following link where `dut-c` is connected with itself ? OR you men something else.  I have some doubt who well meshent is able to do the plumbing here. 

```   - apiVersion: networkop.co.uk/v1beta1
    kind: Topology
    metadata:
        name: dut-c
    spec:
        links:
          - uid: 2001        
            peer_pod: dut-c
            local_intf: e1-15
            peer_intf: e1-16
            ```

Reply:
yeah, I think this is place where it's defined

Reply:
I'd assume that meshnet should be able to create a veth link and move both sides of it into a pod's namespace.

Reply:
@networkop Yes we have an issue with loopback veth link.. i find the links missing

Reply:
@manomugdha, @kingshukdev  Thanks for taking the time to graph the issue out with timeline, it really simplifies it for everyone else and be on same page.

With your suggestion to check the skip per (pod, link) , i m not 100%,  it will resolve the race condition from happening.

Reply:
Pushed a fix for the loopback veth create issue

Reply:
couple of suggestion about "loopback fix"  
Let's not mix it with tie breaker logic (skip, priority etc).  After MAC Vlan creation block (https://github.com/networkop/meshnet-cni/blob/56abbfc48205285cd2b18d5aaf78f69b56c0a7b6/plugin/meshnet.go#L222)  around line 246 do the loopback check, create loopback veth pair and continue. 

I will also suggest to submit "loopback fix" as a separate PR.

Reply:
I agree, I think it would make sense to try and implement that "per-link skip" logic first, see if it fixes the issue with `file exists`  and then do the loopback fix PR.

So about the "per-link skip", does anyone have any suggestions? The most naive way would be to encode device and interface names, separated by a special symbol, e.g. `deviceName:LinkName` and store them in the same place in `status.Skipped` as `[]string`. The other alternative would be to change the API, e.g. 
```
type TopologyStatus struct {
	Skipped     []Skipped `json:"skipped"`
	SrcIP       string   `json:"src_ip"`
	NetNS       string   `json:"net_ns"`
	ContainerID string   `json:"container_id"`
}

type Skipped struct {
  Name    string  `json:"name"`
  Interface string  `json:"intf"`
}
```

wdyt? @sar772004 @kingshukdev @manomugdha

Reply:
> 

i prefer the API change you suggest.. but i m still not convinced it will help the case.. not sure if @manomugdha  and @kingshukdev think about it.

I will open a separate PR for loopback fix and revert it here

Reply:
Using TopologyStatus struct and Skipped struct is a good way.  Minor renaming suggestion :- 
``` 
type Skipped struct {
  PodName    string  `json:"pod_name"`
  Interface string  `json:"intf"`
} 
```
@sar772004 - can you share little more why you think this "per link skip"  information is not enough to resolve it ? In which situation it might fall short ?

Reply:
Reason is, if you see figure @manomugdha  added above, step 4 and 5 can happen still even if we had the per link skip.  
(P.S i still have to go deep into the the skipped logic, so correct me if you think it will work)

Reply:
5 will not happen and will be handled correctly if the skip check is done on individual link basis.  The current skip is set if any single link is not created then we assume the whole node is skipped irrespective of how many links two peer nodes has among them. This is what append at 1, 2. 

For example in your topology there ware 4 links between two peers. When the first link was `not` created, then meshnet marked the  the whole node to be skipped and that caused the problem at  5.

Reply:
does anyone want to try the API change for per-link skipping logic? if not, I can give it a try next week.

Reply:
@networkop , @manomugdha & I can take it up. We are almost done with the “reconciliation” work. It is staged [here ](https://github.com/open-traffic-generator/meshnet-cni/tree/grpc-recon) and currently we are testing it. We can add the skip changes in this branch initially,  keep testing it along with reconciliation and finally submit a combined PR for merge.  We expect to get it done roughly in two weeks.

Reply:
Any update on this ?
Also i notice another failure condition when the pod interfaces are missing

In the cni logs i see this and no veth interfaces were created on "test" pod
time="2023-05-18T21:41:25-04:00" level=error msg="Add: Pod default:test was not a topology pod returning"

Also this message is so misleading
time="2023-05-18T21:41:25-04:00" level=info msg="meshnet cni call successful"

Reply:
Thanks, one question, when the pods are coming up and meshnet links are deployed,
Should the pods be up first and then the meshnet links be deployed? i see some issues with this when pods either join later or if pods are already running and then meshnet links are deployed. Basically not consistent behavior on apply and delete of the pods or meshnet links yaml.

Also, Examples in meshnet repo seem to have the pod spec defined in the same yaml. It might be good to have some tests to separate out the pod bringup and meshnet links bringup. (possibly adding some delay)

Reply:
> Any update on this ? Also i notice another failure condition when the pod interfaces are missing
> 
> In the cni logs i see this and no veth interfaces were created on "test" pod time="2023-05-18T21:41:25-04:00" level=error msg="Add: Pod default:test was not a topology pod returning"
> 
> Also this message is so misleading time="2023-05-18T21:41:25-04:00" level=info msg="meshnet cni call successful"

Ignore this one, i noticed the yaml was having an issue in this case.. (internal bug)

Reply:
> Should the pods be up first and then the meshnet links be deployed?

No. The CR (i.e. links) has to be deployed first. Later when the pods are deployed, K8S invokes meshnet cni. meshnet-cni refers to deployed CR to find out what links it has to create for a pod.  To summarize :- 
1. Deploy the CR
2. Deploy pods.
3. One can remove the deployed pods and bring it back.  
4. Delete the CR (i.e. links yaml) only when one is done with the topology.

Reply:
@sar772004 since PR #80 is merged now, you can play with the networkop/meshnet. No need to use my branch that I will delete anytime.

Reply:
I tried latest meshnet and there is some issue with pods and the links between them, I did not have the bandwidth to check in detail. But its the same setup i mentioned above. Attaching the latest logs and yaml for your reference.
[meshnet-cni.log](https://github.com/networkop/meshnet-cni/files/11580353/meshnet-cni.log)

[meshnet_links.yaml.txt](https://github.com/networkop/meshnet-cni/files/11580367/meshnet_links.yaml.txt)

Reply:
Hi @sar772004, can you please share the corresponding meshnet daemon logs from the node from where you have collected meshnet-cni.log? It is required to see the actual error message why 'Skip' fails.
From the logs, it seems that ultimately failed pods e.g. test, dut-bridge, came up. what are the issues you are observing from application point of view?
I ran this topology both in vxlan and grpc mode. It went fine. In vxlan mode, some interfaces are giving less MTU (1450). so I had to adjust my verification steps. I ran in kind cluster.

Reply:
Hi @manomugdha  Sorry long weekend here, i don't have the setup at the moment, But the issue is when the links are all "veth" . i.e all the pods are in single compute.

All pods had come up eventually. but my use case was failing due to missing synchronization between active/standby CPU pods, some of these backplane links run through dut_bridge pod (acts as a bridge). And the software running on the CPU pod expects the links to be up before it starts.. I do have startup probes on these pods to confirm the backplane link status, but its possible one end of the bridge link was not up.

I will try to set it up again and get the meshnet daemon logs.

NOTE: this worked when i was using #76 itself.